### PR TITLE
statistics: fix unstable store flow

### PR DESCRIPTION
**Pull request by Luffbee**
### What problem does this PR solve? 

Store may report heartbeat much more frequently than the setting frequency, so statistics of stores will become unstable.

### What is changed and how it works?

Use `AvgOverTime` to maintain average over the last setting interval.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: server/statistics/avg_over_time.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	"github.com/phf/go-queue/queue"
+)
+
+type deltaWithInterval struct {
+	delta    float64
+	interval time.Duration
+}
+
+// AvgOverTime maintains change rate in the last interval.
+type AvgOverTime struct {
```
Comment (lhy1024) [inline server/statistics/avg_over_time.go]: I think we need some comments to explain the algorithm

Comment (Luffbee) [reply server/statistics/avg_over_time.go]: Updated.

File: server/statistics/store.go [hunk 2]
```diff
@@ -260,24 +261,25 @@ func (s *StoresStats) GetStoresKeysReadStat() map[uint64]float64 {
 // RollingStoreStats are multiple sets of recent historical records with specified windows size.
 type RollingStoreStats struct {
 	sync.RWMutex
-	bytesWriteRate          MovingAvg
-	bytesReadRate           MovingAvg
-	keysWriteRate           MovingAvg
-	keysReadRate            MovingAvg
+	bytesWriteRate          *AvgOverTime
```
Comment (nolouch) [inline server/statistics/store.go]: Why the region stats also need to change to `aot`?  `aot` can't handle noise.

File: server/statistics/store.go [hunk 3]
```diff
@@ -296,15 +298,15 @@ func collect(records []*pdpb.RecordPair) float64 {
 func (r *RollingStoreStats) Observe(stats *pdpb.StoreStats) {
 	statInterval := stats.GetInterval()
 	interval := statInterval.GetEndTimestamp() - statInterval.GetStartTimestamp()
-	if interval == 0 {
-		return
-	}
+	//if interval == 0 {
```
Comment (nolouch) [inline server/statistics/store.go]: remove.

File: server/statistics/store.go (hunk 2 above)
Comment (Luffbee) [reply server/statistics/store.go]: The noise caused by leader transferring is already handled by TiKV. Other noise I think will mainly come from short interval heartbeats, which will be solved by `AvgOverTime`.

Comment (Luffbee) [reply server/statistics/store.go]: I will revert changes on regions, and do more test. If this works, I will submit another PR.

Comment (nolouch) [reply server/statistics/store.go]: got.

**Comment by nolouch** [comment]
PTAL @lhy1024

File: server/statistics/avg_over_time.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	}
+	return aot.deltaSum / aot.durSum.Seconds()
+}
+
+// Add adds recent change to AvgOverTime.
+func (aot *AvgOverTime) Add(delta float64, interval time.Duration) {
+	aot.que.PushBack(deltaWithInterval{delta, interval})
+	aot.deltaSum += delta
+	aot.durSum += interval
+	if aot.durSum <= aot.avgInterval {
```
Comment (lhy1024) [inline server/statistics/avg_over_time.go]: where do we update `aot.avgInterval`

File: server/statistics/avg_over_time.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+// AvgOverTime maintains change rate in the last avgInterval.
+//
+// AvgOverTime takes changes with their own intervals,
+// stores recent changes that happened in the last avgInterval,
+// then calculates the change rate by (sum of changes) / (sum of intervals).
+type AvgOverTime struct {
+	que         *queue.Queue
+	deltaSum    float64
+	durSum      time.Duration
```
Comment (lhy1024) [inline server/statistics/avg_over_time.go]: how about `duration`? deltaSum and durSum have the same prefix and suffix

File: server/statistics/avg_over_time.go (hunk 4 above)
Comment (Luffbee) [reply server/statistics/avg_over_time.go]: We do not update it.

File: server/statistics/avg_over_time.go (hunk 5 above)
Comment (Luffbee) [reply server/statistics/avg_over_time.go]: I don't think it is a problem.

Comment (lhy1024) [reply server/statistics/avg_over_time.go]: They are similar and make it difficult to read

Comment (Luffbee) [reply server/statistics/avg_over_time.go]: Renamed to `intervalSum`.

**Pull request merged**